File: scripts/enquiry.js
/* Whose desk is this? — /contact.
 * ---------------------------------------------------------------------------
 * The page is complete and correct without this. The bar above the form is in
 * the HTML and says, for every visitor, that this form reaches Kamerun in
 * Douala and Buea. That much needs no JavaScript and gets none.
 *
 * What needs a browser is the specific case: a visitor arriving from the
 * journey builder with ?journey= already written out. That sentence names
 * countries. If none of them is a country Kamerun runs, the visitor is about to
 * send a Cameroonian operator a letter about the Serengeti, and the moment to
 * say so is before the form is filled in, not in the reply a day later.
 *
 * It names countries and operators only. It does not rewrite the enquiry, does
 * not redirect, and does not block sending — a traveller is allowed to write to
 * whoever they like, and being told is the whole of what was missing.
 */
(function () {
  'use strict';

  var node = document.getElementById('fj-reach');
  var bar = document.querySelector('[data-af-handoff]');
  if (!node || !bar) return;

  var reach;
  try { reach = JSON.parse(node.textContent || '{}'); } catch (e) { return; }
  var ours = reach.ours || {};
  var rest = reach.rest || {};
  if (!reach.home || !ours[reach.home]) return;

  /* What the visitor brought with them. The journey sentence is the reliable
     one; the rest of the query string is read too, because a link somebody
     pasted may carry a country and nothing else. */
  var said = '';
  try {
    said = decodeURIComponent((location.search || '').replace(/\+/g, ' '));
  } catch (e) { said = location.search || ''; }
  if (!said) return;

  function saidIn(name) {
    /* Word boundaries by hand: "Chad" must not match inside "Chadian", and a
       country whose name is two words has to match as two words. */
    var safe = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    return new RegExp('(^|[^A-Za-z])' + safe + '([^A-Za-z]|$)', 'i').test(said);
  }

  var named = [], covered = [], slug;
  for (slug in ours) {
    if (saidIn(ours[slug].name)) { named.push(slug); covered.push(slug); }
  }
  for (slug in rest) {
    if (saidIn(rest[slug])) named.push(slug);
  }

  /* Nothing recognisable, or Kamerun's own country is in there: the bar already
     says what it needs to and adding a second sentence would be noise. */
  if (!named.length || covered.indexOf(reach.home) >= 0) return;

  function esc(text) {
    return String(text == null ? '' : text)
      .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
  }

  function listed(parts) {
    if (parts.length === 1) return parts[0];
    return parts.slice(0, -1).join(', ') + ' and ' + parts[parts.length - 1];
  }

  var names = named.map(function (s) {
    return esc((ours[s] || rest[s] || {}).name || rest[s]);
  });

  var say = document.createElement('p');
  say.className = 'fj-from-warn';
  var lead = 'Your journey names ' + listed(names) + '. '
    + esc(ours[reach.home].op) + ' runs ' + esc(ours[reach.home].name)
    + ', so this form is not their desk.';

  if (covered.length) {
    /* One of ours, just not this one — hand them straight over. */
    say.innerHTML = lead + ' ' + covered.map(function (s) {
      return '<a href="' + esc(ours[s].url) + '">' + esc(ours[s].name)
        + ' is run by ' + esc(ours[s].op) + ' &rarr;</a>';
    }).join(' ');
  } else {
    /* None of the three. Say that plainly rather than routing them anywhere:
       the site has already said it does not run a company in the other
       nineteen, and repeating that here is the consistent answer. */
    say.innerHTML = lead
      + ' We do not run a company in ' + listed(names) + ' &mdash; '
      + (named.length > 1 ? 'those countries are' : 'that country is')
      + ' written up here but booked through someone else. Send it anyway and '
      + 'they will say who, or read '
      + listed(named.map(function (s) {
        return '<a href="/portrait/' + esc(s) + '">' + esc((ours[s] || {}).name || rest[s]) + '</a>';
      })) + ' first.';
  }

  bar.appendChild(say);
  bar.setAttribute('data-af-warned', 'true');
}());

File: tools/tourism/enquiry.py
"""Tell a visitor whose enquiry form they are standing in front of.

    python3 tools/tourism/build.py enquiry

/contact is the primary call of the entire site. Every Afrinkong page carries
"Plan a journey" and it points here; so does the closing block on the homepage,
which arrives with ?journey= already written out; so does the 404.

/contact is Kamerun's. It has Kamerun's masthead, Kamerun's four-item nav, a
Douala telephone number, "Plan a circuit" as its call, and a form that opens a
mailto to bonjour@kamerun.cm. Kamerun runs Cameroon. A visitor who has just
built a journey through Kenya and Tanzania on this site lands here, sees a page
that looks like it belongs to the site they were reading, fills in the form,
and emails a Cameroonian operator about the Serengeti.

Nothing on the page tells them. That is the whole defect, and it is not a
design problem — the page is well made — it is that one operator's enquiry desk
is standing in for the group's, silently.

Two things go in, both generated from the dataset so neither can drift:

  a bar under the masthead     saying where this form goes and who reads it,
                               with the way back to the atlas

  a script                     that reads ?journey= (and the URL generally),
                               finds the country names in it, and if Cameroon
                               is not among them says so before the form is
                               filled in rather than after it is sent

No email address is invented here. There is no group enquiry address anywhere
in this project, so the honest move is to name the operator this form reaches
and to point at the countries it does not cover, not to make one up.
"""

import json
import os
import re

from .model import ROOT, load_operators

PAGE = os.path.join(ROOT, "contact.html")
MARKERS = ("handoff", "reach")


def esc(v):
    import html as html_mod
    return html_mod.escape(str(v if v is not None else ""), quote=True)


def covered(countries):
    """-> [(slug, name, operator name, operator url, operator base)]."""
    return [(c.slug, c.name, c.operator.name, c.operator.url, c.operator.base)
            for c in countries if c.operator and c.published]


def block_handoff(countries):
    """The bar under Kamerun's masthead, in Afrinkong's voice."""
    ours = covered(countries)
    # Whoever's desk this page actually is: the operator hosted on this domain
    # rather than on its own. The other two have their own sites and are links
    # away from here, not the page you are standing on.
    here = [row for row in ours if row[3].startswith("/")] or ours[:1]
    who = here[0]
    elsewhere = [row for row in ours if row[0] != who[0]]
    others = "".join(
        '<a href="%s">%s &mdash; %s</a>' % (esc(url), esc(name), esc(op))
        for _slug, name, op, url, _base in elsewhere)
    return (
        '<div class="fj-from" data-af-handoff>\n'
        '  <div class="fj-frame fj-from-in">\n'
        '    <p><b>You are on %s&rsquo;s enquiry page.</b> %s is the Afrinkong '
        'operator for %s, based in %s, and everything sent from this form is '
        'read by them.</p>\n'
        '    <p class="fj-from-else">Somewhere else in Africa? %s or start from '
        '<a href="/atlas">the atlas</a>.</p>\n'
        '  </div>\n'
        '</div>' % (esc(who[2]), esc(who[2]), esc(who[1]), esc(who[4]),
                    others or 'See <a href="/">the other operators</a>'))


def block_reach(countries):
    """What the page knows about which country belongs to whom.

    Inlined rather than fetched: this has to be true before the visitor starts
    typing, and a fetch that has not landed yet is the same as no warning.
    """
    ours = covered(countries)
    home = ours[0][0] if ours else ""
    reach = {
        "home": home,
        "ours": {slug: {"name": name, "op": op, "url": url}
                 for slug, name, op, url, _base in ours},
        "rest": {c.slug: c.name for c in countries
                 if c.published and not c.operator},
    }
    return ('<script type="application/json" id="fj-reach">%s</script>'
            % json.dumps(reach, ensure_ascii=False, separators=(",", ":")))


def splice(src, blocks):
    missing = [n for n in MARKERS if ("<!-- gen:%s -->" % n) not in src]
    if missing:
        raise ValueError("contact.html is missing markers: %s" % ", ".join(missing))
    for name, body in blocks.items():
        pattern = re.compile(r"(<!-- gen:%s -->\n).*?(\s*<!-- /gen:%s -->)"
                             % (name, name), re.S)
        src = pattern.sub(lambda m: m.group(1) + body + m.group(2), src, count=1)
    return src


def run(countries, page=None, log=print):
    page = page or PAGE
    with open(page) as fh:
        src = fh.read()
    live = [c for c in countries if c.published]
    out = splice(src, {"handoff": block_handoff(live), "reach": block_reach(live)})
    changed = out != src
    if changed:
        with open(page, "w") as fh:
            fh.write(out)
    log("%s the enquiry handoff (%d operators, %d countries without one) into %s"
        % ("rewrote" if changed else "no change:",
           sum(1 for c in live if c.operator),
           sum(1 for c in live if not c.operator),
           os.path.relpath(page, ROOT)))
    return changed
